openlibrary/core/observations.py

```diff
@@ -1,26 +1,252 @@
+
+
+ObservationValue = namedtuple('ObservationValue', ['value_id', 'value', 'prev_value_id'])
+
+@cache.memoize(engine="memcache", key="observations", expires=config.get('observation_cache_duration'))
+def get_observations():
+    """
+    Returns a dictionary of observations that are used to populate forms for patron feedback about a book.
+
+    Dictionary has the following structure:
+    {
+        'observations': [
+            {
+                'id': observation_types.id,
+                'label': observation_types.type,
+                'description': observation_types.description,
+                'multi_choice': observation_types.allow_multiple_values,
+                'values': [ observation_values.value list ]
+            }
+        ]
+    }
+
+    return: Dictionary of all possible observations that can be made about a book.
+    """
+    observations = Observations.get_observation_types_and_values()
+
+    observation_dict = {}
+
+    for o in observations:
+        type = o['type']
+
+        if type not in observation_dict:
+            observation_dict[type] = {
+                'id': o['type_id'],
+                'label': type,
+                'description': o['description'],
+                'multi_choice': o['multi_choice'],
+                'values': []
+            }
+
+        observation_dict[type]['values'].insert(0, ObservationValue(o['value_id'], o['value'], o['prev_value']))
+
+    observation_list = []
 
-# URL for TheBestBookOn
-TBBO_URL = config.get('tbbo_url')
+    for v in observation_dict.values():
+        v['values'] = _sort_values(v['values'])
+        # Do not include observation type if it has no values:
+        if len(v['values']):
+            observation_list.append(v)
 
-def post_observation(data, s3_keys):
-    headers = {
-        'x-s3-access': s3_keys['access'],
-        'x-s3-secret': s3_keys['secret']
+    response = {
+        'observations': observation_list
     }
 
-    response = requests.post(TBBO_URL + '/api/observations', data=data, headers=headers)
+    return response
+
+def _sort_values(values_list):
+    """
+    Given a list of one or more ObservationValue tuples, returns a sorted list of observation values.
+
+    ObservationValues are namedtuples with fields ('value_id', 'value', 'prev_value_id').
+    Each field maps to a column in the observation_values table:
+
+    'value_id' -> observation_values.id
+    'value' -> observations_values.value
+    'prev_value_id' -> observation_values.prev_value
+
+    Before values are displayed to patrons in a form, they must first be ordered.  Order is
+    maintained by the observation_values.prev_value column.  Each row has a reference to the
+    ID of the preceding row, or None if the row is the first value in the list.
+
+    return: A sorted list of values.
+    """
+
+    if not len(values_list):
+        return []
+
+    # Add middle list item to sorted list
+    middle_item = values_list.pop(int(len(values_list) / 2))
```

Comment:
```suggestion
    middle_item = values_list.pop(int(values_list) // 2)
```
Use integer division to get an int.

Reply:
Thanks!  I had completely forgotten about integer division...